src/components/main/actionsPanel/navigatorPanel/hooks/useFavicon.ts

```diff
@@ -0,0 +1,109 @@
+import { useContext, useEffect, useRef } from "react";
+import { useDispatch, useSelector } from "react-redux";
+import { RootNodeUid } from "@_constants/main";
+import { THtmlNodeData } from "@_node/html";
+import {
+  ffSelector,
+  MainContext,
+  navigatorSelector,
+  selectFNNode,
+} from "@_redux/main";
+import { TProject } from "@_types/main";
+
+export const useFavicon = (
+  setFaviconFallback: React.Dispatch<React.SetStateAction<boolean>>,
+) => {
+  const dispatch = useDispatch();
+
+  const { file } = useSelector(navigatorSelector);
+  const { selectedItems } = useSelector(ffSelector);
+  const {
+    // navigator
+    workspace,
+    setWorkspace,
+    project,
+    // file tree view
+    ffTree,
+    // node tree view
+    validNodeTree,
+  } = useContext(MainContext);
+
+  const isFirst = useRef(true);
+
+  useEffect(() => {
+    isFirst.current = true;
+  }, [file.uid]);
+
+  useEffect(() => {
```

Comment:
Is it possible to break every segment of code inside the conditions block into a pure function based on what they do.